webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/domain/reports/ExpectedReports.scala

```diff
@@ -199,350 +204,593 @@ final case class NodeConfigIdInfo(
-                            case x                  => Failure(s"Error when parsing the list of components from expected directive report: '${compactRender(x)}'")
-                          }
-          } yield {
-            //if isSystem is not defined => false
-            val isSystem = (json \ "isSystem"  ) match {
-              case JBool(true) => true
-              case _           => false
-            }
+    implicit class _JsonModes7_1(x: NodeModeConfig) {
+      def transform = JsonModes7_1(
+          x.globalPolicyMode.transform
+        , x.nodePolicyMode
+        , x.globalComplianceMode.mode
+        , x.globalComplianceMode.heartbeatPeriod
+        , x.nodeHeartbeatPeriod
+        , x.globalAgentRun.transform
+        , x.nodeAgentRun.map(_.transform)
+      )
+    }
 
-            DirectiveExpectedReports(rid, policyMode(jsonMode), isSystem, components.toList)
-          }
-        case _ =>
-          Failure(s"Error when parsing directive expected reports from json: '${compactRender(json)}'")
-      }
+    final case class JsonPolicy7_1(rid: RuleId, did: DirectiveId) {
+      def transform = PolicyId(rid, did, v0)
+    }
+    implicit class _JsonPolicy7_1(x: PolicyId) {
+      def transform = JsonPolicy7_1(x.ruleId, x.directiveId)
     }
 
-    def value(json : JValue) = {
+    final case class JsonOverrides7_1(
+        p : JsonPolicy7_1
+      , ob: JsonPolicy7_1
+    ) {
+      def transform = OverridenPolicy(p.transform, ob.transform)
+    }
+    implicit class _JsonOverrides7_1(x: OverridenPolicy) {
+      def transform = JsonOverrides7_1(x.policy.transform, x.overridenBy.transform)
+    }
+    final case class JsonExpectedValueId7_1(id: String, v: String) {
+      def transform = ExpectedValueId(v, id)
+    }
+    implicit class _JsonExpectedValueId7_1(x: ExpectedValueId) {
+      def transform = JsonExpectedValueId7_1(x.id, x.value)
+    }
 
-      ( (json \ "value").extractOpt[String]
-      , (json \ "id").extractOpt[String]
-      , (json \ "unexpanded").extractOpt[String]
-      ) match {
-        case ( Some (value),Some(id), _ ) => Full(ExpectedValueId(value,id))
-        case ( Some (value),_, Some(unexpanded) ) => Full(ExpectedValueMatch(value,unexpanded))
-        case _ => Failure("")
+    sealed trait JsonComponentExpectedReport7_1 {
+      def transform: ComponentExpectedReport
+    }
+    implicit class _JsonComponentExpectedReport7_1(x: ComponentExpectedReport) {
+      def transform: JsonComponentExpectedReport7_1 = x match {
+        case a: ValueExpectedReport => a.transform
+        case a: BlockExpectedReport => a.transform
       }
     }
+    final case class JsonValueExpectedReport7_1(
+        vid: String
+      , vs: List[Either[List[String],JsonExpectedValueId7_1]]
+    ) extends JsonComponentExpectedReport7_1 {
+      def transform = ValueExpectedReport(vid, vs.map {
+        case Left(Nil)     => ExpectedValueMatch("None", "None")
```

Comment:
can you add in the comment that the Left is for compatibility with pre-7.1 ? I've been a bit at a lost here